### ISSUE TITLE
Segfault when taking minors over local rings

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1199]
This is an issue that I haven't been able to solve for a long time:
```
restart
needsPackage "LocalRings"

R = ZZ/32003[x_1,x_2,x_3,y_1..y_3, Degrees => {3:{1,0}, 3:{0,1}}]
phi = matrix {{x_1,0,x_3}, {x_2,x_1,0}, {0,x_2,x_1}, {0,0,x_2}}

RP = localRing(R, ideal(x_1,x_2))
psi = sub(phi, RP)

print "first issue"
minors(3, psi)
--error message bumped: attempt to divide by non-unit
--error message bumped: attempt to divide by non-unit
--error message bumped: attempt to divide by non-unit
--o20 = ideal ()
--o20 : Ideal of RP
print "second issue"
minors(3, phi)
-- SIGSEGV
print "done"
```
This seems to be the relevant part of the stack trace:
```
 3# Monomial::is_one() const at /home/mahrud/Projects/M2/M2/M2/BUILD/build/../../Macaulay2/e/monomial.cpp:59
 4# interface_rawMonomialIsOne at /home/mahrud/Projects/M2/M2/M2/Macaulay2/d/interface.dd:135
```

**Comment by DanGrayson** [comment]
Mike, I'm stuck after this:
```
debian$ git diff
diff --git a/M2/Macaulay2/e/det.cpp b/M2/Macaulay2/e/det.cpp
index e30696be5..13ecdfbc7 100644
--- a/M2/Macaulay2/e/det.cpp
+++ b/M2/Macaulay2/e/det.cpp
@@ -201,6 +201,7 @@ ring_elem DetComputation::detmult(ring_elem f1,
   if (!R->is_zero(d))
     {
       ring_elem tmp = R->divide(a, d);  // exact division
+      if (tmp == nullptr) return nullptr;
       R->remove(a);
       a = tmp;
     }
@@ -217,11 +218,17 @@ void DetComputation::gauss(ring_elem **D0,
   ring_elem f = D0[i][pivot_col];
   ring_elem pivot = D0[r][pivot_col];
 
-  for (size_t c = 0; c < pivot_col; c++)
-    D0[i][c] = detmult(pivot, D0[i][c], f, D0[r][c], lastpivot);
-
-  for (size_t c = pivot_col + 1; c <= r; c++)
-    D0[i][c - 1] = detmult(pivot, D0[i][c], f, D0[r][c], lastpivot);
+  for (size_t c = 0; c < pivot_col; c++) {
+    auto ret = detmult(pivot, D0[i][c], f, D0[r][c], lastpivot);
+    if (ret == nullptr) return; // how to indicate an error to the caller?
+    D0[i][c] = ret;
+  }
+
+  for (size_t c = pivot_col + 1; c <= r; c++) {
+    auto ret = detmult(pivot, D0[i][c], f, D0[r][c], lastpivot);
+    if (ret == nullptr) return; // how to indicate an error to the caller?
+    D0[i][c - 1] = ret;
+  }
 
   R->remove(f);
 }
diff --git a/M2/Macaulay2/e/localring.cpp b/M2/Macaulay2/e/localring.cpp
index c3691323f..d2ab84fc0 100644
--- a/M2/Macaulay2/e/localring.cpp
+++ b/M2/Macaulay2/e/localring.cpp
@@ -611,7 +611,7 @@ ring_elem LocalRing::invert(const ring_elem a) const
   if (mRing->is_zero(f->numer) || !is_unit(a))
     {
       ERROR("attempt to invert a non-unit");
-      return zero();
+      return nullptr;
     }
   ring_elem top = mRing->copy(f->denom);
   ring_elem bottom = mRing->copy(f->numer);
@@ -631,7 +631,7 @@ ring_elem LocalRing::divide(const ring_elem a, const ring_elem b) const
   else
     {
       ERROR("attempt to divide by non-unit");
-      return zero();
+      return nullptr;
     }
   return ring_elem(make_elem(top, bottom));
 }
```

**Comment by mikestillman** [comment]
@DanGrayson what is going on?  What is this a diff of?

**Comment by DanGrayson** [comment]
Me starting to try to fix the bugs.

**Comment by mikestillman** [comment]
Fixed in PR #2480.
